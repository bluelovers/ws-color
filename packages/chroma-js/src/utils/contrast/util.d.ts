import { IRGB } from '../../types';
declare const enum EnumContrastFrontThreshold {
    t01 = 131.5,
    t02 = 120
}
export declare function _contrast000(rgba: IRGB): number;
export declare function _contrast001(rgba: IRGB, threshold?: EnumContrastFrontThreshold): boolean;
export declare function _contrast002(rgba: IRGB, threshold?: number): "black" | "white";
export interface IOptionsContrastFront {
    black?: string;
    white?: string;
    threshold?: number;
}
/**
 * get contrast color for use as front text color
 */
export declare function contrastFront(rgba: IRGB, options?: IOptionsContrastFront): string;
export {};
