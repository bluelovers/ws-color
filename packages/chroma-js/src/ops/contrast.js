"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Color_1 = __importDefault(require("../Color"));
const util_1 = require("../utils/contrast/util");
Color_1.default.prototype.contrastFront = function (options) {
    return new Color_1.default(util_1.contrastFront(this.rgba(), options));
};
//# sourceMappingURL=contrast.js.map