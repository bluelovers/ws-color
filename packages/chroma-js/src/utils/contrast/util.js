"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contrastFront = exports._contrast002 = exports._contrast001 = void 0;
function _contrast001(rgba) {
    let [r, g, b] = rgba;
    return (((r * 299) + (g * 587) + (b * 144)) / 1000) >= 131.5;
}
exports._contrast001 = _contrast001;
function _contrast002(rgba) {
    return _contrast001(rgba) ? 'black' : 'white';
}
exports._contrast002 = _contrast002;
/**
 * get contrast color for use as front text color
 */
function contrastFront(rgba, options) {
    var _a;
    let name = _contrast002(rgba);
    return (_a = options === null || options === void 0 ? void 0 : options[name]) !== null && _a !== void 0 ? _a : name;
}
exports.contrastFront = contrastFront;
//# sourceMappingURL=util.js.map