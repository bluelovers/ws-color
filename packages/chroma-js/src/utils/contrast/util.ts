import { IRGB } from '../../types';

const enum EnumContrastFrontThreshold
{
	t01 = 131.5,
	t02 = 120,
}

export function _contrast000(rgba: IRGB)
{
	let [r, g, b] = rgba;

	return (((r * 299) + (g * 587) + (b * 144)) / 1000)
}

export function _contrast001(rgba: IRGB, threshold = EnumContrastFrontThreshold.t02): boolean
{
	return _contrast000(rgba) >= threshold
}

export function _contrast002(rgba: IRGB, threshold?: number)
{
	return _contrast001(rgba, threshold) ? 'black' as const : 'white' as const
}

export interface IOptionsContrastFront
{
	black?: string
	white?: string
	threshold?: number,
}

/**
 * get contrast color for use as front text color
 */
export function contrastFront(rgba: IRGB, options?: IOptionsContrastFront)
{
	let name = _contrast002(rgba)

	return options?.[name] ?? name;
}
