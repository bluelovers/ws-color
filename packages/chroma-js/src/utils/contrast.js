"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Color_1 = __importDefault(require("../Color"));
require("../ops/luminance");
function contrast(a, b) {
    // WCAG contrast ratio
    // see http://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef
    a = new Color_1.default(a);
    b = new Color_1.default(b);
    const l1 = a.luminance();
    const l2 = b.luminance();
    return l1 > l2 ? (l1 + 0.05) / (l2 + 0.05) : (l2 + 0.05) / (l1 + 0.05);
}
exports.default = contrast;
//# sourceMappingURL=contrast.js.map