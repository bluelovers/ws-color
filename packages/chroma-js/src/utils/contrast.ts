import Color from '../Color';
import '../ops/luminance';

function contrast(a: string | Color, b: string | Color)
{
	// WCAG contrast ratio
	// see http://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef
	a = new Color(a);
	b = new Color(b);
	const l1 = a.luminance();
	const l2 = b.luminance();
	return l1 > l2 ? (l1 + 0.05) / (l2 + 0.05) : (l2 + 0.05) / (l1 + 0.05);
}

declare module '../chroma'
{
	interface chroma
	{
		/**
		 * Computes the WCAG contrast ratio between two colors.
		 * A minimum contrast of 4.5:1 is recommended {@link https://www.w3.org/TR/WCAG20-TECHS/G18.html}
		 * to ensure that text is still readable against a background color.
		 */
		contrast: typeof contrast
	}
}

export default contrast
