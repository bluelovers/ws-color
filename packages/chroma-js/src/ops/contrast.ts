import Color from '../Color';
import { _contrast002, contrastFront, IOptionsContrastFront } from '../utils/contrast/util';

declare module '../Color'
{

	interface Color
	{
		/**
		 * get contrast color for use as front text color
		 */
		contrastFront(options?: IOptionsContrastFront): Color;
	}
}

Color.prototype.contrastFront = function (options?: IOptionsContrastFront)
{
	return new Color(contrastFront(this.rgba(), options));
}
