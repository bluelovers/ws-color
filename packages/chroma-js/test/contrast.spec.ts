import { chroma } from '../';
import { _contrast000 } from '../src/utils/contrast/util';

describe(`contrastFront`, () =>
{

	[
		['#906', '#ffffff'],
		['#66F', '#000000'],
		['#800', '#ffffff'],
		['#C60', '#000000'],
		['#EAEA00', '#000000'],
		['#006', '#ffffff'],
		['#360', '#ffffff'],
	].forEach(color => {

		test(color[0], () =>
		{
			let c = chroma(color[0])

			let actual = c.contrastFront();
			let expected = color[1];

			let d = _contrast000(c.rgba())

			console.log(c.hex(), actual.hex(), d)

			console.log(c.named(), actual.named())

			expect(actual.hex()).toStrictEqual(expected)
			expect([c.hex(), actual.hex(), d]).toMatchSnapshot();

		});

	})

})
