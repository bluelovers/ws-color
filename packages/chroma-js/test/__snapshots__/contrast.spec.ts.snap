// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contrastFront #006 1`] = `
[
  "#000066",
  "#ffffff",
  14.688,
]
`;

exports[`contrastFront #66F 1`] = `
[
  "#6666ff",
  "#000000",
  127.092,
]
`;

exports[`contrastFront #360 1`] = `
[
  "#336600",
  "#ffffff",
  75.123,
]
`;

exports[`contrastFront #800 1`] = `
[
  "#880000",
  "#ffffff",
  40.664,
]
`;

exports[`contrastFront #906 1`] = `
[
  "#990066",
  "#ffffff",
  60.435,
]
`;

exports[`contrastFront #C60 1`] = `
[
  "#cc6600",
  "#000000",
  120.87,
]
`;

exports[`contrastFront #EAEA00 1`] = `
[
  "#eaea00",
  "#000000",
  207.324,
]
`;
